### ISSUE TITLE
Allow expression parsing without names

### ISSUE THREAD

**Issue by backbone87** [symfony/symfony#16339]
It would be nice if you could parse expressions without knowing variable names and maybe even function names beforehand.
- Cache any expression in parsed state without knowing its exact evaluation context
- Construct the evaluation context on demand (extremly valueable for expression driven business rules that could use **many** different variables depending on the system complexity)
- Check that you can provide all variables/functions before evaluation and avoid failing evaluations, that may not even be recoverable (fatal error)

The 2nd point can also be achieved by providing proxy objects only exposing very strict sets of your business API and "optimizied for expressions"-methods (lax arg conversion, conveniece accessors, etc.).

**Comment by stof** [comment]
Function are necessary at parse time: they are resolved and validated during the parsing, not during the evaluation. the same is true for variable names, which is necessary to avoid giving access to any variables (which is one of the goal of the component: sandboxing what you can access)

**Comment by backbone87** [comment]
I know how the current parsing process works, thats not the point. 

"giving access to any variables" is only true für compiled expressions, not evaluated ones.
I think providing evaluation and compilation introduce unnecessary complexity into the component. I know the ExpressionLanguage is mainly used for configuration of the DI system, so the compiling is a must have for caching.

But this basically ruins the component for use in "user land" code / business code, because you need more meta information about parsed expressions for better validation and handling of them inside of business rules.

The expression language component with its current feature set is only made to fit the needs of developers when building the software, and is not suited for administrative usage of the resulting software (setting up business rules for shops etc.).

From my point of view, there are multiple "solutions":
- Declare that the expression language component is exclusively made for developer use and every other use is out of scope or has to be done on its own risk or use other libraries. If this is the case, a point could be made, if the expression language deserves to be a standalone component or just a subcomponent of the config/DI system.
- Reevaluate the usage and needed "quality-of-life"-features of the component to be suited for user land application (thats what my ticket is about, but ofc it just points out a few edges, not all of them)
- Provide 2 different components one optimized for developer use and the other for userland application

**Comment by stof** [comment]
> If this is the case, a point could be made, if the expression language deserves to be a standalone component or just a subcomponent of the config/DI system

We already use it in several Symfony components, not only in the DI component, so this is not a question.

and features of the component are the one we needed for the cases where Symfony wanted this component. Removing the current features allowing compilation of expressions would make the component unusable in Symfony in most places using it.

**Comment by dratel29** [comment]
I find myself in the same situations as @backbone87. It would be useful to be able to work with expressions outside of a specific context (for example in a separate admin tool). Checking that no unauthorized properties or functions are access could be done during evaluation instead.

**Comment by backbone87** [comment]
When did i talk about removing current features? I havent talked about any implementation specific concerns yet. I was just pointing out that the current feature set is not suited well for use in business code. If this is desired behavior, i am ok with it. But if the plan is to provide a general purpose tool, i pointed out (some) flaws and (some) missing features based on actual experience (we recently integrated the expression language as a configuration option for specific administrative tasks within our software).

But i can also talk about implementation concerns, if you like:
IMO it is not a concern of the the parser to be aware of variable and function names. The parser should just build the node tree according to the syntax. The used variable and function names are collected while parsing and are part of the parsing result.

There are 2 processes that use a parsed expression:
- Evaluation: The "evaluator" (implicitly through the node tree) already gets passed available functions and variables. It is very easy to compare the provided names with the ones emitted by the parser (and should be part of the parsing result)
- Compilation: The compiler receives functions at construction time and variable names (or name-"translations") at call time. Again the component itself can compare the provided names with the ones emitted by the parser **and** do the "translation" of variable names by itself during compilation.

I have also checked the symfony source and the only place where ExpressionLanguage::parse is called outside of the component itself is in the security bundle when parsing expression for caching.

Also the name check itself can be done without problems in a complete separate component or integration specific when the function and variable names are part of the parsing result. Its as simple as using an array_diff(_key).
